forc-pkg/src/manifest.rs

```diff
@@ -134,3 +138,8 @@ fn default_entry() -> String {
 fn default_url() -> String {
     constants::DEFAULT_NODE_URL.into()
 }
+
+#[test]
+fn test_unused_key() {
+    // TODO
+}
```

Comment:
I'm not sure we can write a useful crate test here to check for warnings without spawning a whole `forc` process, as for now warnings are only emitted via stdout. I think we can remove this for now.

At some point we should do a sweep through `forc` and replace the inline `println`s with a more structured approach ala #885, which should make writing these sorts of tests easier.

Reply:
@eureka-cpu just to see what the warnings look like I would do something like:

1. Add an unused `foo = "blah"` to one of the example project `Forc.toml`'s.
2. Make sure I have my branch's version of `forc` installed. You can do this with `cargo install --path ./forc`
3. Run `forc build` on the example I modified.
4. Copy all the output and paste it here :) 

Hope this helps!

Reply:
Helps massively, thank you!

Reply:
Okay, so there's a few things: 

1. this is the output when I add an unused key `[[foo]]`, as intended I get:  `WARNING! unused manifest key: foo`, however I get that warning twice. Once at the beginning of the build, and once at the end before `Compiled script "address_test".`
2. if you were to add @mitchmindtree 's example `foo = "blah"` you actually get the `not implemented` error from `forc-pkg/src/pkg.rs:696:45`, but instead of it telling you what wasn't implemented it just shows where the panic occured.

Reply:
>  I think what's happening is that it's passing the warning for each library that gets compiled, which makes cargo's reasoning for all the extra code make sense.

Adding a `BTreeSet` removes the duplicates.

Reply:
> ... however I get that warning twice.

Ah this is likely due to `Manifest::validate` method getting called within the `Manifest` constructor, and the constructor getting called multiple times during a `forc build`. Ideally the `Manifest` should only be constructed once per package, but its convenient for some of the graph related code to construct it on demand. This would be easier to address with #885. For now I think it's OK to keep the duplicates (2 warnings is better than none, and its the uncommon case), but we should add a note about addressing this to #885.

> if you were to add @mitchmindtree 's example foo = "blah" you actually get the not implemented error

It sounds like you might have put this line under the `[dependencies]` section of the and its hitting the "Simple" dependency declaration which assumes the string to be a semver version. As you discovered this is `unimplemented!` but, rather than panicking we should return a descriptive error. I'll do a separate PR for this now.

Out of curiosity, do you get a warning if you place `foo = "blah"` under the `[project]` heading? Or does it perhaps result in a deserialization error?

Reply:
>Ah this is likely due to Manifest::validate method getting called within the Manifest constructor, and the constructor getting called multiple times during a forc build.

I added a `dbg!` to check when we were calling it and it seemed to me that the manifest was being read a second time in `BuildPlan`, though I could be wrong. I added this to the issue I filed #928 .

>Out of curiosity, do you get a warning if you place foo = "blah" under the [project] heading? Or does it perhaps result in a deserialization error?

This actually doesn't even get acknowledged... No warning or error...

Reply:
forc build
  WARNING! unused manifest key: project.foo
  Compiled library "core".
warning
  --> 247270e7d8911e754a989b2fe280c462605ae2f6/src/token.sw:64:20
   |
62 | 
63 |         // if an ouput is found of type `OutputVariable`:
64 |         if type == OUTPUT_VARIABLE_TYPE {
   |                    -------------------- This cast, from integer type of width sixty four to integer type of width eight, will lose precision.
65 |             let amount = asm(n: index, a, amount_ptr, output: output_start) {
66 |                 addi amount_ptr output i40;
   |
warning
  --> /Users/killswitchengagebiggestfan/.forc/git/checkouts/std-ebb222698a8d2787/247270e7d8911e754a989b2fe280c462605ae2f6/src/chain:7:1
   |
 5 |   
 6 |   
 7 | / pub enum AuthError {
 8 | |     ContextError: (),
 9 | | }
   | |_- This declaration is never used.
10 |   
11 |   pub enum Sender {
   |
  Compiled library "std" with 2 warnings.
  WARNING! unused manifest key: project.foo
  Compiled script "address_test".
  Bytecode size is 1824 bytes.

Reply:
Also, if we add `foo = "blah"` twice, instead of getting an error we just get a second warning for `unused manifest key: project.foo`